src/sentry/models/outbox.py

```diff
@@ -270,29 +270,33 @@ def send_signal(self) -> None:
     def drain_shard(
         self, flush_all: bool = False, _test_processing_barrier: threading.Barrier | None = None
     ) -> None:
-        # When we are flushing in a local context, we don't care about outboxes created concurrently --
-        # at best our logic depends on previously created outboxes.
-        latest_shard_row: OutboxBase | None = None
-        if not flush_all:
-            latest_shard_row = self.selected_messages_in_shard().last()
-            # If we're not flushing all possible shards, and we don't see any immediately values,
-            # drop.
-            if latest_shard_row is None:
-                return
-
-        shard_row: OutboxBase | None
-        while True:
-            with self.process_shard(latest_shard_row) as shard_row:
-                if shard_row is None:
-                    break
+        def _innner() -> None:
```

Comment:
This unfolds the previous `on_commit` logic from `OutboxBase.save` into the explicit drain_shard call.
Essentially, `drain_shard` calls itself are no longer synchronous to the call site, but synchronous to completing the outer transactions, which is more correct and prevents some test bugs where SiloModes were mixing in SPLIT_DB_MODE due to the way `using=` can change via router in that mdoe based on SiloModes.

Reply:
Why three n's?

Reply:
> Essentially, drain_shard calls itself are no longer synchronous to the call site, but synchronous to completing the outer transactions

Are there any situations where we truly want the synchronous form of drain? I could see implementing the sync form for when we need to roll back work within a transaction; however, I also get that draining a shard within a transaction has the potential to be highly problematic or to have severe performance impact.